<a href="https://colab.research.google.com/github/ujeongmin/ESAA/blob/main/ESAA_HW_4_12.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

(파이썬 머신러닝 완벽 가이드)
## [CH8 텍스트 분석]



### 07. 문서 군집화 소개와 실습(Opinion Review 데이터 세트)
#### 문서 군집화 개념
비슷한 텍스트 구성의 문서를 군집화하는 것  
텍스트 분류 기반의 문서 분류는 사전에 결정 카테고리 값을 가진 학습 데이터 세트가 필요한 데 반해, 문서 군집화는 학습 데이터 세트가 필요 없는 비지도학습 기반으로 동작

#### Opinion Review 데이터 세트를 이용한 문서 군집화 수행하기




In [ ]:
import pandas as pd
import glob, os
from google.colab import drive
drive.mount('/content/gdrive')

path = r'/content/gdrive/MyDrive/ESAA/data/topics2'
all_files = glob.glob(os.path.join(path, '*.data'))
filename_list = []
opinion_text = []

for file_ in all_files:
  df = pd.read_table(file_, index_col=None, header=0, encoding='latin1')
  filename_ = file_.split('\\')[-1]
  filename = filename_.split('.')[0]
  filename_list.append(filename)
  opinion_text.append(df.to_string())

document_df = pd.DataFrame({'filename':filename_list, 'opinion_text':opinion_text})
document_df.head()

Mounted at /content/gdrive


filename  \
0  /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
1  /content/gdrive/MyDrive/ESAA/data/topics2/accu...   
2  /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
3  /content/gdrive/MyDrive/ESAA/data/topics2/dire...   
4  /content/gdrive/MyDrive/ESAA/data/topics2/disp...   

                                        opinion_text  
0                                                ...  
1                                                ...  
2                                                ...  
3                                                ...  
4                                                ...

* 각 파일 이름 자체만으로 의견의 텍스트가 어떠한 제품/서비스에 대한 리뷰인지 알 수 있음
* 문서를 TF-IDF 피처 벡터화

In [ ]:
nltk.download('all')

In [ ]:
# 앞에서 정의한 함수 가져오기

from nltk.stem import WordNetLemmatizer
import nltk
import string

remove_punct_dict = dict((ord(punct), None) for punct in string.punctuation)
lemmar = WordNetLemmatizer()

def LemTokens(tokens):
    return [lemmar.lemmatize(token) for token in tokens]

def LemNormalize(text):
    return LemTokens(nltk.word_tokenize(text.lower().translate(remove_punct_dict)))

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer(tokenizer=LemNormalize, stop_words='english',
                             ngram_range=(1,2), min_df=0.05, max_df=0.85)
# opinion_text 칼럼 값으로 피처 벡터화 수행
feature_vect = tfidf_vect.fit_transform(document_df['opinion_text'])

/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:409: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['ha', 'le', 'u', 'wa'] not in stop_words.
  warnings.warn(


* K-평균을 적용

In [ ]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화 수행
km_cluster = KMeans(n_clusters=5, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


In [ ]:
document_df['cluster_label'] = cluster_label
document_df.head()

filename  \
0  /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
1  /content/gdrive/MyDrive/ESAA/data/topics2/accu...   
2  /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
3  /content/gdrive/MyDrive/ESAA/data/topics2/dire...   
4  /content/gdrive/MyDrive/ESAA/data/topics2/disp...   

                                        opinion_text  cluster_label  
0                                                ...              1  
1                                                ...              0  
2                                                ...              1  
3                                                ...              0  
4                                                ...              0

* 파일명으로 정렬해서 군집화 결과 살펴보기

In [ ]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
1   /content/gdrive/MyDrive/ESAA/data/topics2/accu...   
3   /content/gdrive/MyDrive/ESAA/data/topics2/dire...   
4   /content/gdrive/MyDrive/ESAA/data/topics2/disp...   
7   /content/gdrive/MyDrive/ESAA/data/topics2/sate...   
8   /content/gdrive/MyDrive/ESAA/data/topics2/scre...   
13  /content/gdrive/MyDrive/ESAA/data/topics2/spee...   
15  /content/gdrive/MyDrive/ESAA/data/topics2/upda...   
17  /content/gdrive/MyDrive/ESAA/data/topics2/voic...   

                                         opinion_text  cluster_label  
1                                                 ...              0  
3                                                 ...              0  
4                                                 ...              0  
7                                                 ...              0  
8                                                 ...              0  
13                                                ...              0  
15                                                ...              0  
17                                                ...              0

In [ ]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
18  /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
0   /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
2   /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
6   /content/gdrive/MyDrive/ESAA/data/topics2/keyb...   
29  /content/gdrive/MyDrive/ESAA/data/topics2/perf...   
9   /content/gdrive/MyDrive/ESAA/data/topics2/scre...   
10  /content/gdrive/MyDrive/ESAA/data/topics2/scre...   
11  /content/gdrive/MyDrive/ESAA/data/topics2/size...   
12  /content/gdrive/MyDrive/ESAA/data/topics2/soun...   
16  /content/gdrive/MyDrive/ESAA/data/topics2/vide...   

                                         opinion_text  cluster_label  
18                                                ...              1  
0                                                 ...              1  
2                                                 ...              1  
6                                                 ...              1  
29                                                ...              1  
9                                                 ...              1  
10                                                ...              1  
11                                                ...              1  
12      headphone jack i got a clear case for it a...              1  
16                                                ...              1

In [ ]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
37  /content/gdrive/MyDrive/ESAA/data/topics2/bath...   
23  /content/gdrive/MyDrive/ESAA/data/topics2/food...   
24  /content/gdrive/MyDrive/ESAA/data/topics2/food...   
39  /content/gdrive/MyDrive/ESAA/data/topics2/free...   
41  /content/gdrive/MyDrive/ESAA/data/topics2/loca...   
27  /content/gdrive/MyDrive/ESAA/data/topics2/loca...   
43  /content/gdrive/MyDrive/ESAA/data/topics2/park...   
31  /content/gdrive/MyDrive/ESAA/data/topics2/pric...   
32  /content/gdrive/MyDrive/ESAA/data/topics2/room...   
46  /content/gdrive/MyDrive/ESAA/data/topics2/room...   
33  /content/gdrive/MyDrive/ESAA/data/topics2/room...   
48  /content/gdrive/MyDrive/ESAA/data/topics2/serv...   
34  /content/gdrive/MyDrive/ESAA/data/topics2/serv...   
35  /content/gdrive/MyDrive/ESAA/data/topics2/serv...   
49  /content/gdrive/MyDrive/ESAA/data/topics2/staf...   
36  /content/gdrive/MyDrive/ESAA/data/topics2/staf...   

                                         opinion_text  cluster_label  
37                                                ...              2  
23                                                ...              2  
24                                                ...              2  
39                                                ...              2  
41                                                ...              2  
27                                                ...              2  
43                                                ...              2  
31                                                ...              2  
32                                                ...              2  
46                                                ...              2  
33                                                ...              2  
48                                                ...              2  
34                                                ...              2  
35                                                ...              2  
49                                                ...              2  
36                                                ...              2

In [ ]:
document_df[document_df['cluster_label']==3].sort_values(by='filename')

filename  \
19  /content/gdrive/MyDrive/ESAA/data/topics2/butt...   
21  /content/gdrive/MyDrive/ESAA/data/topics2/eyes...   
5   /content/gdrive/MyDrive/ESAA/data/topics2/feat...   
22  /content/gdrive/MyDrive/ESAA/data/topics2/font...   
28  /content/gdrive/MyDrive/ESAA/data/topics2/navi...   
30  /content/gdrive/MyDrive/ESAA/data/topics2/pric...   
14  /content/gdrive/MyDrive/ESAA/data/topics2/spee...   

                                         opinion_text  cluster_label  
19                                                ...              3  
21                                                ...              3  
5                                                 ...              3  
22                                                ...              3  
28                                                ...              3  
30                                                ...              3  
14                                                ...              3

In [ ]:
document_df[document_df['cluster_label']==4].sort_values(by='filename')

filename  \
38  /content/gdrive/MyDrive/ESAA/data/topics2/comf...   
20  /content/gdrive/MyDrive/ESAA/data/topics2/comf...   
25  /content/gdrive/MyDrive/ESAA/data/topics2/gas_...   
40  /content/gdrive/MyDrive/ESAA/data/topics2/inte...   
26  /content/gdrive/MyDrive/ESAA/data/topics2/inte...   
42  /content/gdrive/MyDrive/ESAA/data/topics2/mile...   
44  /content/gdrive/MyDrive/ESAA/data/topics2/perf...   
45  /content/gdrive/MyDrive/ESAA/data/topics2/qual...   
47  /content/gdrive/MyDrive/ESAA/data/topics2/seat...   
50  /content/gdrive/MyDrive/ESAA/data/topics2/tran...   

                                         opinion_text  cluster_label  
38                                                ...              4  
20                                                ...              4  
25                                                ...              4  
40                                                ...              4  
26                                                ...              4  
42                                                ...              4  
44                                                ...              4  
45                                                ...              4  
47                                                ...              4  
50                                                ...              4

* 파일명에 따라 주제를 살펴보면 군집 개수가 약간 많아 세분화되어 군집화된 경향이 있음, 3개 군집으로 낮추기

In [ ]:
# 3개 집합으로 군집화 수행
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

# 소속 군집을 cluster_label 칼럼으로 할당하고 cluster_label 값으로 정렬
document_df['cluster_label'] = cluster_label
document_df.sort_values(by='cluster_label')

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


filename  \
0   /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
30  /content/gdrive/MyDrive/ESAA/data/topics2/pric...   
29  /content/gdrive/MyDrive/ESAA/data/topics2/perf...   
28  /content/gdrive/MyDrive/ESAA/data/topics2/navi...   
22  /content/gdrive/MyDrive/ESAA/data/topics2/font...   
19  /content/gdrive/MyDrive/ESAA/data/topics2/butt...   
18  /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
17  /content/gdrive/MyDrive/ESAA/data/topics2/voic...   
16  /content/gdrive/MyDrive/ESAA/data/topics2/vide...   
15  /content/gdrive/MyDrive/ESAA/data/topics2/upda...   
14  /content/gdrive/MyDrive/ESAA/data/topics2/spee...   
13  /content/gdrive/MyDrive/ESAA/data/topics2/spee...   
21  /content/gdrive/MyDrive/ESAA/data/topics2/eyes...   
11  /content/gdrive/MyDrive/ESAA/data/topics2/size...   
1   /content/gdrive/MyDrive/ESAA/data/topics2/accu...   
2   /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
12  /content/gdrive/MyDrive/ESAA/data/topics2/soun...   
4   /content/gdrive/MyDrive/ESAA/data/topics2/disp...   
5   /content/gdrive/MyDrive/ESAA/data/topics2/feat...   
3   /content/gdrive/MyDrive/ESAA/data/topics2/dire...   
7   /content/gdrive/MyDrive/ESAA/data/topics2/sate...   
8   /content/gdrive/MyDrive/ESAA/data/topics2/scre...   
9   /content/gdrive/MyDrive/ESAA/data/topics2/scre...   
10  /content/gdrive/MyDrive/ESAA/data/topics2/scre...   
6   /content/gdrive/MyDrive/ESAA/data/topics2/keyb...   
48  /content/gdrive/MyDrive/ESAA/data/topics2/serv...   
41  /content/gdrive/MyDrive/ESAA/data/topics2/loca...   
39  /content/gdrive/MyDrive/ESAA/data/topics2/free...   
46  /content/gdrive/MyDrive/ESAA/data/topics2/room...   
37  /content/gdrive/MyDrive/ESAA/data/topics2/bath...   
36  /content/gdrive/MyDrive/ESAA/data/topics2/staf...   
35  /content/gdrive/MyDrive/ESAA/data/topics2/serv...   
34  /content/gdrive/MyDrive/ESAA/data/topics2/serv...   
31  /content/gdrive/MyDrive/ESAA/data/topics2/pric...   
32  /content/gdrive/MyDrive/ESAA/data/topics2/room...   
27  /content/gdrive/MyDrive/ESAA/data/topics2/loca...   
49  /content/gdrive/MyDrive/ESAA/data/topics2/staf...   
24  /content/gdrive/MyDrive/ESAA/data/topics2/food...   
23  /content/gdrive/MyDrive/ESAA/data/topics2/food...   
33  /content/gdrive/MyDrive/ESAA/data/topics2/room...   
43  /content/gdrive/MyDrive/ESAA/data/topics2/park...   
47  /content/gdrive/MyDrive/ESAA/data/topics2/seat...   
45  /content/gdrive/MyDrive/ESAA/data/topics2/qual...   
44  /content/gdrive/MyDrive/ESAA/data/topics2/perf...   
25  /content/gdrive/MyDrive/ESAA/data/topics2/gas_...   
40  /content/gdrive/MyDrive/ESAA/data/topics2/inte...   
38  /content/gdrive/MyDrive/ESAA/data/topics2/comf...   
26  /content/gdrive/MyDrive/ESAA/data/topics2/inte...   
20  /content/gdrive/MyDrive/ESAA/data/topics2/comf...   
42  /content/gdrive/MyDrive/ESAA/data/topics2/mile...   
50  /content/gdrive/MyDrive/ESAA/data/topics2/tran...   

                                         opinion_text  cluster_label  
0                                                 ...              0  
30                                                ...              0  
29                                                ...              0  
28                                                ...              0  
22                                                ...              0  
19                                                ...              0  
18                                                ...              0  
17                                                ...              0  
16                                                ...              0  
15                                                ...              0  
14                                                ...              0  
13                                                ...              0  
21                                                ...              0  
11                                                ...              0  
1              

In [ ]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
1   /content/gdrive/MyDrive/ESAA/data/topics2/accu...   
18  /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
0   /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
2   /content/gdrive/MyDrive/ESAA/data/topics2/batt...   
19  /content/gdrive/MyDrive/ESAA/data/topics2/butt...   
3   /content/gdrive/MyDrive/ESAA/data/topics2/dire...   
4   /content/gdrive/MyDrive/ESAA/data/topics2/disp...   
21  /content/gdrive/MyDrive/ESAA/data/topics2/eyes...   
5   /content/gdrive/MyDrive/ESAA/data/topics2/feat...   
22  /content/gdrive/MyDrive/ESAA/data/topics2/font...   
6   /content/gdrive/MyDrive/ESAA/data/topics2/keyb...   
28  /content/gdrive/MyDrive/ESAA/data/topics2/navi...   
29  /content/gdrive/MyDrive/ESAA/data/topics2/perf...   
30  /content/gdrive/MyDrive/ESAA/data/topics2/pric...   
7   /content/gdrive/MyDrive/ESAA/data/topics2/sate...   
8   /content/gdrive/MyDrive/ESAA/data/topics2/scre...   
9   /content/gdrive/MyDrive/ESAA/data/topics2/scre...   
10  /content/gdrive/MyDrive/ESAA/data/topics2/scre...   
11  /content/gdrive/MyDrive/ESAA/data/topics2/size...   
12  /content/gdrive/MyDrive/ESAA/data/topics2/soun...   
13  /content/gdrive/MyDrive/ESAA/data/topics2/spee...   
14  /content/gdrive/MyDrive/ESAA/data/topics2/spee...   
15  /content/gdrive/MyDrive/ESAA/data/topics2/upda...   
16  /content/gdrive/MyDrive/ESAA/data/topics2/vide...   
17  /content/gdrive/MyDrive/ESAA/data/topics2/voic...   

                                         opinion_text  cluster_label  
1                                                 ...              0  
18                                                ...              0  
0                                                 ...              0  
2                                                 ...              0  
19                                                ...              0  
3                                                 ...              0  
4                                                 ...              0  
21                                                ...              0  
5                                                 ...              0  
22                                                ...              0  
6                                                 ...              0  
28                                                ...              0  
29                                                ...              0  
30                                                ...              0  
7                                                 ...              0  
8                                                 ...              0  
9                                                 ...              0  
10                                                ...              0  
11                                                ...              0  
12      headphone jack i got a clear case for it a...              0  
13                                                ...              0  
14                                                ...              0  
15                                                ...              0  
16                                                ...              0  
17                                                ...              0

In [ ]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
37  /content/gdrive/MyDrive/ESAA/data/topics2/bath...   
23  /content/gdrive/MyDrive/ESAA/data/topics2/food...   
24  /content/gdrive/MyDrive/ESAA/data/topics2/food...   
39  /content/gdrive/MyDrive/ESAA/data/topics2/free...   
41  /content/gdrive/MyDrive/ESAA/data/topics2/loca...   
27  /content/gdrive/MyDrive/ESAA/data/topics2/loca...   
43  /content/gdrive/MyDrive/ESAA/data/topics2/park...   
31  /content/gdrive/MyDrive/ESAA/data/topics2/pric...   
32  /content/gdrive/MyDrive/ESAA/data/topics2/room...   
46  /content/gdrive/MyDrive/ESAA/data/topics2/room...   
33  /content/gdrive/MyDrive/ESAA/data/topics2/room...   
48  /content/gdrive/MyDrive/ESAA/data/topics2/serv...   
34  /content/gdrive/MyDrive/ESAA/data/topics2/serv...   
35  /content/gdrive/MyDrive/ESAA/data/topics2/serv...   
49  /content/gdrive/MyDrive/ESAA/data/topics2/staf...   
36  /content/gdrive/MyDrive/ESAA/data/topics2/staf...   

                                         opinion_text  cluster_label  
37                                                ...              1  
23                                                ...              1  
24                                                ...              1  
39                                                ...              1  
41                                                ...              1  
27                                                ...              1  
43                                                ...              1  
31                                                ...              1  
32                                                ...              1  
46                                                ...              1  
33                                                ...              1  
48                                                ...              1  
34                                                ...              1  
35                                                ...              1  
49                                                ...              1  
36                                                ...              1

In [ ]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
38  /content/gdrive/MyDrive/ESAA/data/topics2/comf...   
20  /content/gdrive/MyDrive/ESAA/data/topics2/comf...   
25  /content/gdrive/MyDrive/ESAA/data/topics2/gas_...   
40  /content/gdrive/MyDrive/ESAA/data/topics2/inte...   
26  /content/gdrive/MyDrive/ESAA/data/topics2/inte...   
42  /content/gdrive/MyDrive/ESAA/data/topics2/mile...   
44  /content/gdrive/MyDrive/ESAA/data/topics2/perf...   
45  /content/gdrive/MyDrive/ESAA/data/topics2/qual...   
47  /content/gdrive/MyDrive/ESAA/data/topics2/seat...   
50  /content/gdrive/MyDrive/ESAA/data/topics2/tran...   

                                         opinion_text  cluster_label  
38                                                ...              2  
20                                                ...              2  
25                                                ...              2  
40                                                ...              2  
26                                                ...              2  
42                                                ...              2  
44                                                ...              2  
45                                                ...              2  
47                                                ...              2  
50                                                ...              2

* Cluster #0은 포터블 전자기기 리뷰, #1은 호텔 리뷰, #2는 자동차 리뷰로 군집화가 잘 구성

#### 군집별 핵심 단어 추출하기
cluster_centers_는 배열 값으로 제공되며, 행은 개별 군집 열은 개별 피처를 의미  
각 배열 내의 값은 개별 군집내의 상대 위치를 숫자 값으로 표현한 일종의 좌표 값

In [ ]:
cluster_centers = km_cluster.cluster_centers_
print('cluster_centers shape :', cluster_centers.shape)
print(cluster_centers)

cluster_centers shape : (3, 4611)
[[0.01005322 0.         0.         ... 0.00706287 0.         0.        ]
 [0.         0.00099499 0.00174637 ... 0.         0.00183397 0.00144581]
 [0.         0.00092551 0.         ... 0.         0.         0.        ]]


* 군집이 3개, word 피처가 4611개로 구성되어있음, 1에 가까울수록 중심과 가까운 값


* argsort()[:,::-1] 이용하면 배열 내 값이 큰 순으로 정렬된 위치 인덱스 값을 반환 (정렬된 값이 아니라 인덱스)
* get_cluster_details() 함수를 생성해 각 군집별 핵심 단어 찾기

In [ ]:
# 군집별 top n 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명을 반환함
def get_cluster_details(cluster_model, cluster_data, feature_names,
                        clusters_num, top_n_features=10):
  cluster_details = {}

  # cluster_centers array의 값이 큰 순으로 정렬된 인덱스 값을 반환
  # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함
  centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:,::-1]

  # 개별 군집별로 반복하면서 핵심 단어, 그 단어의 중심 위치 상댓값, 대상 파일명 입력
  for cluster_num in range(clusters_num):
    # 개별 군집별 정보를 담을 데이터 초기화
    cluster_details[cluster_num] = {}
    cluster_details[cluster_num]['cluster'] = cluster_num

    # cluster_centers_.argsort()[:, ::-1]로 구한 인덱스를 이용해 top n 피처 단어를 구함
    top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
    top_features = [feature_names[ind] for ind in top_feature_indexes]

    # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함
    top_feature_values = cluster_model.cluster_centers_[cluster_num,
                                                        top_feature_indexes].tolist()

    # cluster_details 딕셔너리 객체에 개별 군집별 핵심단어와 중심위치 상댓값, 해당 파일명 입력
    cluster_details[cluster_num]['top_features'] = top_features
    cluster_details[cluster_num]['top_features_value'] = top_feature_values
    filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
    filenames = filenames.values.tolist()

    cluster_details[cluster_num]['filenames'] = filenames

  return cluster_details

* get_cluster_details()를 호출하면 딕셔너리를 원소로 가지는 리스트인 cluster_details를 반환
* 개별 군집번호, 핵심 단어, 핵심단어 중심 위치 상댓값, 파일명 속성 값 정보가 들어있는데 보기 좋게 출력하기 위해 별도의 print_cluster_details() 함수 생성

In [ ]:
def print_cluster_details(cluster_details):
  for cluster_num, cluster_detail in cluster_details.items():
    print('###### Cluster {0}'.format(cluster_num))
    print('Top features :', cluster_detail['top_features'])
    print('Reviews 파일명 :', cluster_detail['filenames'][:7])
    print('==============================================')

In [ ]:
feature_names = tfidf_vect.get_feature_names_out()

cluster_details = get_cluster_details(cluster_model=km_cluster, cluster_data=document_df,
                                      feature_names=feature_names, clusters_num=3, top_n_features=10)
print_cluster_details(cluster_details)

###### Cluster 0
Top features : ['screen', 'battery', 'keyboard', 'battery life', 'life', 'kindle', 'direction', 'video', 'size', 'voice']
Reviews 파일명 : ['/content/gdrive/MyDrive/ESAA/data/topics2/battery-life_ipod_nano_8gb', '/content/gdrive/MyDrive/ESAA/data/topics2/accuracy_garmin_nuvi_255W_gps', '/content/gdrive/MyDrive/ESAA/data/topics2/battery-life_netbook_1005ha', '/content/gdrive/MyDrive/ESAA/data/topics2/directions_garmin_nuvi_255W_gps', '/content/gdrive/MyDrive/ESAA/data/topics2/display_garmin_nuvi_255W_gps', '/content/gdrive/MyDrive/ESAA/data/topics2/features_windows7', '/content/gdrive/MyDrive/ESAA/data/topics2/keyboard_netbook_1005ha']
###### Cluster 1
Top features : ['room', 'hotel', 'service', 'staff', 'food', 'location', 'bathroom', 'clean', 'price', 'parking']
Reviews 파일명 : ['/content/gdrive/MyDrive/ESAA/data/topics2/food_holiday_inn_london', '/content/gdrive/MyDrive/ESAA/data/topics2/food_swissotel_chicago', '/content/gdrive/MyDrive/ESAA/data/topics2/location_holiday_

* Cluster #0에서는 화면과 배터리 수명 등이, #1에서는 방과 서비스 등이, #2에서는 실내 인테리어, 좌석, 연료 효율 등이 핵심 단어로 군집화됨